docs/rfcs/component-universal-telemetry.md

```diff
@@ -0,0 +1,212 @@
+Similarly, the `memory_limiter` processor is a singleton, so its telemetry is not specific to a pipeline.
+
+## Auto-Instrumentation Mechanism
+
+The mechanism of telemetry capture should be _external_ to components. Specifically, we should observe telemetry at each point where a
+component passes data to another component, and, at each point where a component consumes data from another component. In terms of the
+component graph, every _edge_ in the graph will have two layers of instrumentation - one for the producing component and one for the
+consuming component. Importantly, each layer generates telemetry ascribed to a single component instance, so by having two layers per
+edge we can describe both sides of each handoff independently.
+
+Telemetry captured by this mechanism should be associated with an instrumentation scope corresponding to the package which implements
+the mechanism. Currently, that package is `service/internal/graph`, but this may change in the future. Notably, this telemetry is not
+ascribed to individual component packages, both because the instrumentation scope is intended to describe the origin of the telemetry,
+and because no mechanism is presently identified which would allow us to determine the characteristics of a component-specific scope.
+
+### Auto-Instrumented Metrics
+
+There are two straightforward measurements that can be made on any pdata:
+
+1. A count of "items" (spans, data points, or log records). These are low cost but broadly useful, so they should be enabled by default.
+2. A measure of size, based on [ProtoMarshaler.Sizer()](https://github.com/open-telemetry/opentelemetry-collector/blob/9907ba50df0d5853c34d2962cf21da42e15a560d/pdata/ptrace/pb.go#L11).
+  These are high cost to compute, so by default they should be disabled (and not calculated).
+
+The location of these measurements can be described in terms of whether the data is "consumed" or "produced", from the perspective of the
+component to which the telemetry is attributed. Metrics which contain the term "produced" describe data which is emitted from the component,
+while metrics which contain the term "consumed" describe data which is received by the component.
+
+For both metrics, an `outcome` attribute with possible values `success` and `failure` should be automatically recorded, corresponding to
+whether or not the corresponding function call returned an error. Specifically, consumed measurements will be recorded with `outcome` as
+`failure` when a call from the previous component the `ConsumeX` function returns an error, and `success` otherwise. Likewise, produced
+measurements will be recorded with `outcome` as `failure` when a call to the next consumer's `ConsumeX` function returns an error, and
+`success` otherwise.
+
+```yaml
+    otelcol.receiver.produced.items:
+      enabled: true
+      description: Number of items emitted from the receiver.
+      unit: "{items}"
```

Comment:
Unit should not be plural according to https://opentelemetry.io/docs/specs/semconv/general/metrics/#instrument-units.

Also, after reading the RFC, it's still unclear to me why signal goes to an attribute and unit is an item now. According to https://opentelemetry.io/docs/specs/semconv/general/metrics/#general-guidelines aggregation over all signals should give a meaningful value. I guess potentially it can be meaningful but unlikely useful. Is there any other reason to not keep separate metric names for spans, logs records and metric points other than reaggregation capability?

Reply:
I will update the units.

> it's still unclear to me why signal goes to an attribute and unit is an item now.

Let's separate this into two questions:

### 1. Should there be a `signal` attribute?

Yes, because these attributes are not just for metrics. With metrics if we put the signal in the metric name that may be enough, but you cannot do the same for e.g. a log. If we want to be able to correlate all telemetry from each instance, then we should have the exact same attributes across all signals.

### 2. Should signal type be in metric names? (in which case we use units like "data points" instead of "items")

We could do this, but if we agree with the rationale for question 1, then it would be redundant information. If only looking at this point, maybe it could go either way, but I think the question of aggregation is more important.

> I guess potentially it can be meaningful but unlikely useful.

I think the total size of data is very meaningful if aggregated across signals, as this tends to correlate to a large extent with volume of data flowing over the network and cost to customers. I agree item count is less obvious though. However, if the size case makes sense, then I think it is helpful to keep the structure of the two metrics the same.

--- 

Therefore, I think the best approach is as proposed in the RFC. This allows us correlate across all data types while being specific enough to ascribe telemetry to individual component instances. It also allows us to aggregate across telemetry types. And although not terribly important, it does not contain redundant information.

Reply:
I've updated the units to `item` instead of `items`

Reply:
A couple of examples that use the approach on this RFC: Vector [uses `component_received_events_counts`](https://vector.dev/docs/administration/monitoring/) and Fluentbit [uses `fluentbit_input_records_total`](https://docs.fluentbit.io/manual/administration/monitoring#telemetry-pipeline). I think we should go with the approach currently in the RFC

Reply:
Thanks for the detailed response, @djaglowski. That makes sense to me. Let's keep the current proposal. Given that, I think we should better document what an `item` is somewhere in the docs maybe.